File: src/main/java/io/vertx/mqtt/messages/MqttUnsubscribeMessage.java
/*
 * Copyright 2016 Red Hat Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.vertx.mqtt.messages;

import io.netty.handler.codec.mqtt.MqttProperties;
import io.vertx.codegen.annotations.CacheReturn;
import io.vertx.codegen.annotations.GenIgnore;
import io.vertx.codegen.annotations.VertxGen;
import io.vertx.mqtt.messages.impl.MqttUnsubscribeMessageImpl;

import java.util.List;

/**
 * Represents an MQTT UNSUBSCRIBE message
 */
@VertxGen
public interface MqttUnsubscribeMessage extends MqttMessage {

  /**
   * Create a concrete instance of a Vert.x unsubscribe message
   *
   * @param messageId message identifier
   * @param topics    list of topics to unsubscribe
   */
  @GenIgnore
  static MqttUnsubscribeMessage create(int messageId, List<String> topics) {

    return new MqttUnsubscribeMessageImpl(messageId, topics, MqttProperties.NO_PROPERTIES);
  }

  /**
   * Create a concrete instance of a Vert.x unsubscribe message
   *
   * @param messageId message identifier
   * @param topics    list of topics to unsubscribe
   * @param properties UNSUBSCRIBE message properties
   */
  @GenIgnore(GenIgnore.PERMITTED_TYPE)
  static MqttUnsubscribeMessage create(int messageId, List<String> topics, MqttProperties properties) {

    return new MqttUnsubscribeMessageImpl(messageId, topics, properties);
  }


  /**
   * @return  List of topics to unsubscribe
   */
  @CacheReturn
  List<String> topics();

  /**
   * @return MQTT properties
   */
  @CacheReturn
  @GenIgnore(GenIgnore.PERMITTED_TYPE)
  MqttProperties properties();
}

File: src/main/resources/vertx-mqtt-server-js/mqtt_unsubscribe_message.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module vertx-mqtt-server-js/mqtt_unsubscribe_message */
var utils = require('vertx-js/util/utils');
var MqttMessage = require('vertx-mqtt-server-js/mqtt_message');

var io = Packages.io;
var JsonObject = io.vertx.core.json.JsonObject;
var JMqttUnsubscribeMessage = io.vertx.mqtt.messages.MqttUnsubscribeMessage;

/**
 Represents an MQTT UNSUBSCRIBE message

 @class
*/
var MqttUnsubscribeMessage = function(j_val) {

  var j_mqttUnsubscribeMessage = j_val;
  var that = this;
  MqttMessage.call(this, j_val);

  /**

   @public

   @return {number} Message identifier
   */
  this.messageId = function() {
    var __args = arguments;
    if (__args.length === 0) {
      if (that.cachedmessageId == null) {
        that.cachedmessageId = j_mqttUnsubscribeMessage["messageId()"]();
      }
      return that.cachedmessageId;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public

   @return {Array.<string>} List of topics to unsubscribe
   */
  this.topics = function() {
    var __args = arguments;
    if (__args.length === 0) {
      if (that.cachedtopics == null) {
        that.cachedtopics = j_mqttUnsubscribeMessage["topics()"]();
      }
      return that.cachedtopics;
    } else throw new TypeError('function invoked with invalid arguments');
  };

  // A reference to the underlying Java delegate
  // NOTE! This is an internal API and must not be used in user code.
  // If you rely on this property your code is likely to break if we change it / remove it without warning.
  this._jdel = j_mqttUnsubscribeMessage;
};

MqttUnsubscribeMessage._jclass = utils.getJavaClass("io.vertx.mqtt.messages.MqttUnsubscribeMessage");
MqttUnsubscribeMessage._jtype = {
  accept: function(obj) {
    return MqttUnsubscribeMessage._jclass.isInstance(obj._jdel);
  },
  wrap: function(jdel) {
    var obj = Object.create(MqttUnsubscribeMessage.prototype, {});
    MqttUnsubscribeMessage.apply(obj, arguments);
    return obj;
  },
  unwrap: function(obj) {
    return obj._jdel;
  }
};
MqttUnsubscribeMessage._create = function(jdel) {
  var obj = Object.create(MqttUnsubscribeMessage.prototype, {});
  MqttUnsubscribeMessage.apply(obj, arguments);
  return obj;
}
module.exports = MqttUnsubscribeMessage;